nexus/db-queries/src/db/datastore/sled.rs

```diff
@@ -43,11 +48,244 @@ use omicron_uuid_kinds::GenericUuid;
+fn pick_sled_reservation_target(
+    log: &Logger,
+    mut targets: HashSet<SledUuid>,
+    anti_affinity_sleds: Vec<(AffinityPolicy, Uuid)>,
+    affinity_sleds: Vec<(AffinityPolicy, Uuid)>,
+) -> Result<SledUuid, SledReservationError> {
+    let (banned, mut unpreferred): (HashSet<_>, HashSet<_>) =
+        anti_affinity_sleds.into_iter().partition_map(|(policy, id)| {
+            let id = SledUuid::from_untyped_uuid(id);
+            match policy {
+                AffinityPolicy::Fail => Either::Left(id),
+                AffinityPolicy::Allow => Either::Right(id),
+            }
+        });
+    let (required, mut preferred): (HashSet<_>, HashSet<_>) =
+        affinity_sleds.into_iter().partition_map(|(policy, id)| {
+            let id = SledUuid::from_untyped_uuid(id);
+            match policy {
+                AffinityPolicy::Fail => Either::Left(id),
+                AffinityPolicy::Allow => Either::Right(id),
+            }
+        });
+
+    if !banned.is_empty() {
+        info!(
+            log,
+            "anti-affinity policy prohibits placement on {} sleds", banned.len();
+            "banned" => ?banned,
+        );
+    }
+    if !required.is_empty() {
+        info!(
+            log,
+            "affinity policy requires placement on {} sleds", required.len();
+            "required" => ?required,
+        );
+    }
+
+    if required.len() > 1 {
+        return Err(SledReservationError::TooManyAffinityConstraints);
+    }
+    if let Some(required_id) = required.iter().next() {
+        // If we have a "required" sled, it must be chosen.
+        if banned.contains(&required_id) {
+            return Err(
+                SledReservationError::ConflictingAntiAndAffinityConstraints,
+            );
+        }
+        if !targets.contains(&required_id) {
+            return Err(SledReservationError::RequiredAffinitySledNotValid);
+        }
+        return Ok(*required_id);
+    }
+
+    // We have no "required" sleds, but might have preferences.
+    targets = targets.difference(&banned).cloned().collect();
+
+    // Only consider "preferred" sleds that are viable targets
+    preferred = targets.intersection(&preferred).cloned().collect();
+    // Only consider "unpreferred" sleds that are viable targets
+    unpreferred = targets.intersection(&unpreferred).cloned().collect();
+
+    // If a target is both preferred and unpreferred, it is removed
+    // from both sets.
+    let both = preferred.intersection(&unpreferred).cloned().collect();
+    preferred = preferred.difference(&both).cloned().collect();
+    unpreferred = unpreferred.difference(&both).cloned().collect();
+
+    if let Some(target) = preferred.iter().next() {
+        return Ok(*target);
+    }
+    targets = targets.difference(&unpreferred).cloned().collect();
+    if let Some(target) = targets.iter().next() {
+        return Ok(*target);
+    }
+    if let Some(target) = unpreferred.iter().next() {
+        return Ok(*target);
+    }
```

Comment:
somewhat nitpicky and probably unimportant, but it occurs to me that we could do slightly less `collect`ing here:
```suggestion
    if let Some(target) = preferred.difference(&both).cloned().next() {
        return Ok(*target);
    }
    
    unpreferred = unpreferred.difference(&both).cloned().collect();
    targets = targets.difference(&unpreferred).cloned().collect();
    if let Some(target) = targets.iter().next() {
        return Ok(*target);
    }
    if let Some(target) = unpreferred.iter().next() {
        return Ok(*target);
    }
```

This way, we avoid allocating a new `HashSet` for `preferred`, as we never actually need it to be a set, we just need the first value from the `difference` iterator, and we only allocate the new `uinpreferred` and `targets` if we need them because `preferred` was empty.

Reply:
Sounds good, updated to use this one-fewer-collect approach!